homeassistant/components/alarmdecoder.py

```diff
@@ -116,15 +112,33 @@ def stop_alarmdecoder(event):
     @callback
     def handle_message(sender, message):
         """Handle message from AlarmDecoder."""
-        async_dispatcher_send(hass, SIGNAL_PANEL_MESSAGE, message)
+        dispatcher_send(hass, SIGNAL_PANEL_MESSAGE, message)
 
     def zone_fault_callback(sender, zone):
         """Handle zone fault from AlarmDecoder."""
-        async_dispatcher_send(hass, SIGNAL_ZONE_FAULT, zone)
+        dispatcher_send(hass, SIGNAL_ZONE_FAULT, zone)
 
     def zone_restore_callback(sender, zone):
         """Handle zone restore from AlarmDecoder."""
-        async_dispatcher_send(hass, SIGNAL_ZONE_RESTORE, zone)
+        dispatcher_send(hass, SIGNAL_ZONE_RESTORE, zone)
+
+    @callback
+    def handle_connect():
+        """Execute connection in event loop."""
```

Comment:
Why does this need to be run within the event loop? This callback will be called when `controller.open` is called at the end of `setup` here in this module, so I don't see any reason why we should not just let the same thread do all the job.

I think you should remove the callback annotation of this callback and switch to sync versions of all the calls inside this callback.

Reply:
I believe the way the alarmdecoder works is that controller.open sends an 'open' command to the alarmdecoder thread which then calls on_open within that thread (not the current one).  I don't think we want to block the callback thread during platform setup, so I figured we should push it back to the event loop to handle.

Reply:
I just dug into the alarmdecoder code a bit further and it appears you are right.  There is no reason for the on_open callback since the open function happens serially.

Reply:
I don't think that's correct. There's only a reader thread used in the alarmdecoder that is separate. The call to `on_open` via controller.open` is all done in the same thread, ie the same thread as `setup` here is run from.
https://github.com/nutechsoftware/alarmdecoder/blob/master/alarmdecoder/devices/serial_device.py#L124